url_matcher/matcher.py

```diff
@@ -150,11 +161,16 @@ def get(self, identifier: Any) -> Optional[Patterns]:
         return self.patterns.get(identifier)
 
     def match(self, url: str) -> Optional[Any]:
+        return next(self.match_all(url), None)
+
+    def match_all(self, url: str) -> Iterator[Any]:
         domain = get_domain(url)
+        unique: Set[Any] = set()
         for matcher in chain(self.matchers_by_domain.get(domain) or [], self.matchers_by_domain.get("") or []):
             if matcher.match(url):
-                return matcher.identifier
-        return None
+                if matcher.identifier not in unique:
```

Comment:
I might be missing something but doesn't URL Matcher guarantee uniqueness in its identifier because of how the `.add_or_update()` method currently works? wherein if an existing identifier is provided, it would simply update the old URL Pattern with the new one.

Perhaps it will be more clear if we have a test for this.

Reply:
The only test that fails without this change is `test_matcher_single_rule[Match product lists in books.toscrape.com]`, `match_all()` returns the identifier 3 times there. That's because there are 3 include patterns in that test case. Is this the expected behavior?

Reply:
Aha, I get it now. So if we have more than 1 URL Pattern when adding it to the URLMatcher, like this:

```python
from url_matcher import Patterns, URLMatcher 

matcher = URLMatcher()
matcher.add_or_update(23, Patterns(['example.com/products', 'example.com/category']))
matcher.matchers_by_domain
```

It'd result in a repeated `PatternsMatcher` instance:

```python
{
  'example.com': [
    PatternsMatcher(
      identifier=23,
      patterns=Patterns(
        include=('example.com/products', 'example.com/category'),
        exclude=(),
        priority=500
      ),
      include_matchers=[
        <url_matcher.patterns.PatternMatcher object at 0x103e57c10>,
        <url_matcher.patterns.PatternMatcher object at 0x103e57d50>],
      exclude_matchers=[]
    ),
    PatternsMatcher(
      identifier=23,
      patterns=Patterns(
        include=('example.com/products', 'example.com/category'),
        exclude=(),
        priority=500
      ),
      include_matchers=[
        <url_matcher.patterns.PatternMatcher object at 0x103e57c10>,
        <url_matcher.patterns.PatternMatcher object at 0x103e57d50>],
      exclude_matchers=[]
    )
  ]
}
```

Not sure why this is the case but it doesn't look efficient.

Reply:
Created https://github.com/zytedata/url-matcher/issues/10#issue-2217863565 to track this.